# First Python Notebook

Hello world!

In [99]:
import pandas as pd

In [100]:
my_list = [2, 4, 6, 8]

In [101]:
my_series = pd.Series(my_list)

In [102]:
my_series.sum()

20

In [103]:
my_series.max()

8

In [104]:
my_series.min()

2

In [105]:
my_series.mean()

5.0

In [106]:
my_series.median()

5.0

In [107]:
my_series.std()

2.581988897471611

In [108]:
my_series.describe()

count    4.000000
mean     5.000000
std      2.581989
min      2.000000
25%      3.500000
50%      5.000000
75%      6.500000
max      8.000000
dtype: float64

In [109]:
my_list2 = [2, 4, 6, 8, 10]

In [110]:
my_series2 = pd.Series(my_list2)

In [111]:
my_series2.describe()

count     5.000000
mean      6.000000
std       3.162278
min       2.000000
25%       4.000000
50%       6.000000
75%       8.000000
max      10.000000
dtype: float64

In [112]:
my_series2.median()

6.0

In [113]:
my_series2.mean()

6.0

In [114]:
csv_file ='http://www.firstpythonnotebook.org/_static/committees.csv'

In [115]:
committees = pd.read_csv('committees.csv', sep = ',')

In [116]:
committees

ocd_prop_id  calaccess_prop_id  \
0    ocd-contest/b51dc64d-3562-4913-a190-69f5088c22a6            1376258   
1    ocd-contest/b51dc64d-3562-4913-a190-69f5088c22a6            1376258   
2    ocd-contest/b51dc64d-3562-4913-a190-69f5088c22a6            1376258   
3    ocd-contest/b51dc64d-3562-4913-a190-69f5088c22a6            1376258   
4    ocd-contest/85990193-9d6f-4600-b8e7-bf1317841d82            1362198   
5    ocd-contest/85990193-9d6f-4600-b8e7-bf1317841d82            1362198   
6    ocd-contest/85990193-9d6f-4600-b8e7-bf1317841d82            1362198   
7    ocd-contest/9fb6a199-7520-433c-a290-83e016a9d389            1376142   
8    ocd-contest/9fb6a199-7520-433c-a290-83e016a9d389            1376142   
9    ocd-contest/9fb6a199-7520-433c-a290-83e016a9d389            1376142   
10   ocd-contest/9fb6a199-7520-433c-a290-83e016a9d389            1376142   
11   ocd-contest/e0c5f9f0-63a9-4c15-9a5c-017e20ac15bc            1381642   
12   ocd-contest/e0c5f9f0-63a9-4c15-9a5c-017e20ac15bc            1381642   
13   ocd-contest/e0c5f9f0-63a9-4c15-9a5c-017e20ac15bc            1381642   
14   ocd-contest/e0c5f9f0-63a9-4c15-9a5c-017e20ac15bc            1381642   
15   ocd-contest/446ec3da-7de1-45f0-9f97-228e53c6a9f6            1382647   
16   ocd-contest/446ec3da-7de1-45f0-9f97-228e53c6a9f6            1382647   
17   ocd-contest/446ec3da-7de1-45f0-9f97-228e53c6a9f6            1382647   
18   ocd-contest/446ec3da-7de1-45f0-9f97-228e53c6a9f6            1382647   
19   ocd-contest/446ec3da-7de1-45f0-9f97-228e53c6a9f6            1382647   
20   ocd-contest/446ec3da-7de1-45f0-9f97-228e53c6a9f6            1382647   
21   ocd-contest/446ec3da-7de1-45f0-9f97-228e53c6a9f6            1382647   
22   ocd-contest/446ec3da-7de1-45f0-9f97-228e53c6a9f6            1382647   
23   ocd-contest/c2efe0d3-e9f4-4dfd-a0b7-a35757ef4ae9            1381640   
24   ocd-contest/c2efe0d3-e9f4-4dfd-a0b7-a35757ef4ae9            1381640   
25   ocd-contest/c2efe0d3-e9f4-4dfd-a0b7-a35757ef4ae9            1381640   
26   ocd-contest/c2efe0d3-e9f4-4dfd-a0b7-a35757ef4ae9            1381640   
27   ocd-contest/c2efe0d3-e9f4-4dfd-a0b7-a35757ef4ae9            1381640   
28   ocd-contest/c2efe0d3-e9f4-4dfd-a0b7-a35757ef4ae9            1381640   
29   ocd-contest/c2efe0d3-e9f4-4dfd-a0b7-a35757ef4ae9            1381640   
..                                                ...                ...   
72   ocd-contest/e02dc48a-2ed3-4a87-859b-26c90f0539e7            1381803   
73   ocd-contest/e02dc48a-2ed3-4a87-859b-26c90f0539e7            1381803   
74   ocd-contest/232cb72c-1b11-4293-b4e2-a181e80d172e            1381868   
75   ocd-contest/232cb72c-1b11-4293-b4e2-a181e80d172e            1381868   
76   ocd-contest/232cb72c-1b11-4293-b4e2-a181e80d172e            1381868   
77   ocd-contest/232cb72c-1b11-4293-b4e2-a181e80d172e            1381868   
78   ocd-contest/232cb72c-1b11-4293-b4e2-a181e80d172e            1381868   
79   ocd-contest/232cb72c-1b11-4293-b4e2-a181e80d172e            1381868   
80   ocd-contest/232cb72c-1b11-4293-b4e2-a181e80d172e            1381868   
81   ocd-contest/232cb72c-1b11-4293-b4e2-a181e80d172e            1381868   
82   ocd-contest/232cb72c-1b11-4293-b4e2-a181e80d172e            1381868   
83   ocd-contest/232cb72c-1b11-4293-b4e2-a181e80d172e            1381868   
84   ocd-contest/232cb72c-1b11-4293-b4e2-a181e80d172e            1381868   
85   ocd-contest/cac74f20-f59f-4857-8750-5f832a3549c9            1381520   
86   ocd-contest/19eca91d-e494-4ee0-a345-8f96eeec8a3e            1381724   
87   ocd-contest/19eca91d-e494-4ee0-a345-8f96eeec8a3e            1381724   
88   ocd-contest/19eca91d-e494-4ee0-a345-8f96eeec8a3e            1381724   
89   ocd-contest/19eca91d-e494-4ee0-a345-8f96eeec8a3e            1381724   
90   ocd-contest/19eca91d-e494-4ee0-a345-8f96eeec8a3e            1381724   
91   ocd-contest/19eca91d-e494-4ee0-a345-8f96eeec8a3e            1381724   
92   ocd-contest/19eca91d-e494-4ee0-a345-8f96eeec8a3e            1381724   
93   ocd-conte

In [117]:
committees.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 102 entries, 0 to 101
Data columns (total 8 columns):
ocd_prop_id               102 non-null object
calaccess_prop_id         102 non-null int64
ccdc_prop_id              102 non-null int64
prop_name                 102 non-null object
ccdc_committee_id         102 non-null int64
calaccess_committee_id    102 non-null int64
committee_name            102 non-null object
committee_position        102 non-null object
dtypes: int64(4), object(4)
memory usage: 6.5+ KB


In [118]:
contrib = pd.read_csv('contributions.csv', sep = ',')

In [119]:
contrib

calaccess_committee_id  \
0                     1386560   
1                     1386560   
2                     1386560   
3                     1386560   
4                     1386560   
5                     1386560   
6                     1386560   
7                     1386560   
8                     1386560   
9                     1386560   
10                    1386560   
11                    1386560   
12                    1386560   
13                    1386560   
14                    1386560   
15                    1386560   
16                    1386560   
17                    1386560   
18                    1386560   
19                    1386560   
20                    1386560   
21                    1386560   
22                    1386560   
23                    1386560   
24                    1386560   
25                    1386560   
26                    1372902   
27                    1372902   
28                    1372902   
29                    1372902   
...                       ...   
56349                 1374469   
56350                 1374469   
56351                 1374469   
56352                 1374469   
56353                 1374469   
56354                 1374469   
56355                 1374469   
56356                 1374469   
56357                 1374469   
56358                 1374469   
56359                 1374469   
56360                 1374469   
56361                 1374469   
56362                 1374469   
56363                 1374469   
56364                 1374469   
56365                 1374469   
56366                 1374469   
56367                 1362973   
56368                 1362973   
56369                 1362973   
56370                 1362973   
56371                 1362973   
56372                 1362973   
56373                 1362973   
56374                 1362973   
56375                 1362973   
56376                 1362973   
56377                 1362973   
56378                 1362973   

                                          committee_name  calaccess_filing_id  \
0             ADULT USE CAMPAIGN FOR PROPOSITION 64; THE              2083796   
1             ADULT USE CAMPAIGN FOR PROPOSITION 64; THE              2083796   
2             ADULT USE CAMPAIGN FOR PROPOSITION 64; THE              2083796   
3             ADULT USE CAMPAIGN FOR PROPOSITION 64; THE              2083796   
4             ADULT USE CAMPAIGN FOR PROPOSITION 64; THE              2083796   
5             ADULT USE CAMPAIGN FOR PROPOSITION 64; THE              2083796   
6             ADULT USE CAMPAIGN FOR PROPOSITION 64; THE              2083796   
7             ADULT USE CAMPAIGN FOR PROPOSITION 64; THE              2083796   
8             ADULT USE CAMPAIGN FOR PROPOSITION 64; THE              2083796   
9             ADULT USE CAMPAIGN FOR PROPOSITION 64; THE              2083796   
10            ADULT USE CAMPAIGN FOR PROPOSITION 64; THE              2083796   
11            ADULT USE CAMPAIGN FOR PROPOSITION 64; THE              2083796   
12            ADULT USE CAMPAIGN FOR PROPOSITION 64; THE              2083796   
13            ADULT USE CAMPAIGN FOR PROPOSITION 64; THE              2083796   
14            ADULT USE CAMPAIGN FOR PROPOSITION 64; THE              2083796   
15            ADULT USE CAMPAIGN FOR PROPOSITION 64; THE              2083796   
16            ADULT USE CAMPAIGN FOR PROPOSITION 64; THE              2083796   
17            ADULT USE CAMPAIGN FOR PROPOSITION 64; THE              2098758   
18            ADULT USE CAMPAIGN FOR PROPOSITION 64; THE              2098758   
19            ADULT USE CAMPAIGN FOR PROPOSITION 64; THE              2098758   
20            ADULT USE CAMPAIGN FOR PROPOSITION 64; THE              2131665   
21            ADULT USE CAMPAIGN FOR PROPOSITION 64; THE              2131665   
22            ADULT USE CAMPAIGN FOR PROPOSITION 64; THE              2131665   
23            AD

In [120]:
contrib.head()

calaccess_committee_id                              committee_name  \
0                 1386560  ADULT USE CAMPAIGN FOR PROPOSITION 64; THE   
1                 1386560  ADULT USE CAMPAIGN FOR PROPOSITION 64; THE   
2                 1386560  ADULT USE CAMPAIGN FOR PROPOSITION 64; THE   
3                 1386560  ADULT USE CAMPAIGN FOR PROPOSITION 64; THE   
4                 1386560  ADULT USE CAMPAIGN FOR PROPOSITION 64; THE   

   calaccess_filing_id date_received contributor_lastname  \
0              2083796    2016-09-18              BERGMAN   
1              2083796    2016-09-18                KAHLE   
2              2083796    2016-07-15             MCDEVITT   
3              2083796    2016-08-10      WARREN-SAMARIPA   
4              2083796    2016-07-26      WARREN-SAMARIPA   

  contributor_firstname contributor_city contributor_state contributor_zip  \
0              GRETCHEN    SPRING VALLEY                CA           91978   
1                 MYRNA        SAN DIEGO                CA           92109   
2                   LEO        ESCONDIDO                CA           92025   
3             STEPHANIE        SAN DIEGO                CA           92116   
4             STEPHANIE        SAN DIEGO                CA           92116   

        contributor_employer contributor_occupation  \
0                 A NEW PATH     EXECUTIVE DIRECTOR   
1   NATIONAL SCHOOL DISTRICT                TEACHER   
2              LIFE IONIZERS    SEO/CONTENT MANAGER   
3  STEPHANIE WARREN SAMARIPA           ENTREPRENEUR   
4  STEPHANIE WARREN SAMARIPA           ENTREPRENEUR   

   contributor_is_self_employed  amount  
0                         False    84.0  
1                         False    35.0  
2                         False   198.0  
3                         False   -50.0  
4                          True    50.0

In [121]:
contrib.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 56379 entries, 0 to 56378
Data columns (total 13 columns):
calaccess_committee_id          56379 non-null int64
committee_name                  56379 non-null object
calaccess_filing_id             56379 non-null int64
date_received                   56379 non-null object
contributor_lastname            56375 non-null object
contributor_firstname           53411 non-null object
contributor_city                56369 non-null object
contributor_state               56363 non-null object
contributor_zip                 56366 non-null object
contributor_employer            48572 non-null object
contributor_occupation          53273 non-null object
contributor_is_self_employed    56379 non-null bool
amount                          56379 non-null float64
dtypes: bool(1), float64(1), int64(2), object(9)
memory usage: 5.2+ MB


In [122]:
props = committees

In [123]:
props

ocd_prop_id  calaccess_prop_id  \
0    ocd-contest/b51dc64d-3562-4913-a190-69f5088c22a6            1376258   
1    ocd-contest/b51dc64d-3562-4913-a190-69f5088c22a6            1376258   
2    ocd-contest/b51dc64d-3562-4913-a190-69f5088c22a6            1376258   
3    ocd-contest/b51dc64d-3562-4913-a190-69f5088c22a6            1376258   
4    ocd-contest/85990193-9d6f-4600-b8e7-bf1317841d82            1362198   
5    ocd-contest/85990193-9d6f-4600-b8e7-bf1317841d82            1362198   
6    ocd-contest/85990193-9d6f-4600-b8e7-bf1317841d82            1362198   
7    ocd-contest/9fb6a199-7520-433c-a290-83e016a9d389            1376142   
8    ocd-contest/9fb6a199-7520-433c-a290-83e016a9d389            1376142   
9    ocd-contest/9fb6a199-7520-433c-a290-83e016a9d389            1376142   
10   ocd-contest/9fb6a199-7520-433c-a290-83e016a9d389            1376142   
11   ocd-contest/e0c5f9f0-63a9-4c15-9a5c-017e20ac15bc            1381642   
12   ocd-contest/e0c5f9f0-63a9-4c15-9a5c-017e20ac15bc            1381642   
13   ocd-contest/e0c5f9f0-63a9-4c15-9a5c-017e20ac15bc            1381642   
14   ocd-contest/e0c5f9f0-63a9-4c15-9a5c-017e20ac15bc            1381642   
15   ocd-contest/446ec3da-7de1-45f0-9f97-228e53c6a9f6            1382647   
16   ocd-contest/446ec3da-7de1-45f0-9f97-228e53c6a9f6            1382647   
17   ocd-contest/446ec3da-7de1-45f0-9f97-228e53c6a9f6            1382647   
18   ocd-contest/446ec3da-7de1-45f0-9f97-228e53c6a9f6            1382647   
19   ocd-contest/446ec3da-7de1-45f0-9f97-228e53c6a9f6            1382647   
20   ocd-contest/446ec3da-7de1-45f0-9f97-228e53c6a9f6            1382647   
21   ocd-contest/446ec3da-7de1-45f0-9f97-228e53c6a9f6            1382647   
22   ocd-contest/446ec3da-7de1-45f0-9f97-228e53c6a9f6            1382647   
23   ocd-contest/c2efe0d3-e9f4-4dfd-a0b7-a35757ef4ae9            1381640   
24   ocd-contest/c2efe0d3-e9f4-4dfd-a0b7-a35757ef4ae9            1381640   
25   ocd-contest/c2efe0d3-e9f4-4dfd-a0b7-a35757ef4ae9            1381640   
26   ocd-contest/c2efe0d3-e9f4-4dfd-a0b7-a35757ef4ae9            1381640   
27   ocd-contest/c2efe0d3-e9f4-4dfd-a0b7-a35757ef4ae9            1381640   
28   ocd-contest/c2efe0d3-e9f4-4dfd-a0b7-a35757ef4ae9            1381640   
29   ocd-contest/c2efe0d3-e9f4-4dfd-a0b7-a35757ef4ae9            1381640   
..                                                ...                ...   
72   ocd-contest/e02dc48a-2ed3-4a87-859b-26c90f0539e7            1381803   
73   ocd-contest/e02dc48a-2ed3-4a87-859b-26c90f0539e7            1381803   
74   ocd-contest/232cb72c-1b11-4293-b4e2-a181e80d172e            1381868   
75   ocd-contest/232cb72c-1b11-4293-b4e2-a181e80d172e            1381868   
76   ocd-contest/232cb72c-1b11-4293-b4e2-a181e80d172e            1381868   
77   ocd-contest/232cb72c-1b11-4293-b4e2-a181e80d172e            1381868   
78   ocd-contest/232cb72c-1b11-4293-b4e2-a181e80d172e            1381868   
79   ocd-contest/232cb72c-1b11-4293-b4e2-a181e80d172e            1381868   
80   ocd-contest/232cb72c-1b11-4293-b4e2-a181e80d172e            1381868   
81   ocd-contest/232cb72c-1b11-4293-b4e2-a181e80d172e            1381868   
82   ocd-contest/232cb72c-1b11-4293-b4e2-a181e80d172e            1381868   
83   ocd-contest/232cb72c-1b11-4293-b4e2-a181e80d172e            1381868   
84   ocd-contest/232cb72c-1b11-4293-b4e2-a181e80d172e            1381868   
85   ocd-contest/cac74f20-f59f-4857-8750-5f832a3549c9            1381520   
86   ocd-contest/19eca91d-e494-4ee0-a345-8f96eeec8a3e            1381724   
87   ocd-contest/19eca91d-e494-4ee0-a345-8f96eeec8a3e            1381724   
88   ocd-contest/19eca91d-e494-4ee0-a345-8f96eeec8a3e            1381724   
89   ocd-contest/19eca91d-e494-4ee0-a345-8f96eeec8a3e            1381724   
90   ocd-contest/19eca91d-e494-4ee0-a345-8f96eeec8a3e            1381724   
91   ocd-contest/19eca91d-e494-4ee0-a345-8f96eeec8a3e            1381724   
92   ocd-contest/19eca91d-e494-4ee0-a345-8f96eeec8a3e            1381724   
93   ocd-conte

In [124]:
props.prop_name

0      PROPOSITION 051 - SCHOOL BONDS. FUNDING FOR K-...
1      PROPOSITION 051 - SCHOOL BONDS. FUNDING FOR K-...
2      PROPOSITION 051 - SCHOOL BONDS. FUNDING FOR K-...
3      PROPOSITION 051 - SCHOOL BONDS. FUNDING FOR K-...
4      PROPOSITION 052 - STATE FEES ON HOSPITALS. FED...
5      PROPOSITION 052 - STATE FEES ON HOSPITALS. FED...
6      PROPOSITION 052 - STATE FEES ON HOSPITALS. FED...
7      PROPOSITION 053 - REVENUE BONDS. STATEWIDE VOT...
8      PROPOSITION 053 - REVENUE BONDS. STATEWIDE VOT...
9      PROPOSITION 053 - REVENUE BONDS. STATEWIDE VOT...
10     PROPOSITION 053 - REVENUE BONDS. STATEWIDE VOT...
11     PROPOSITION 054 - LEGISLATURE. LEGISLATION AND...
12     PROPOSITION 054 - LEGISLATURE. LEGISLATION AND...
13     PROPOSITION 054 - LEGISLATURE. LEGISLATION AND...
14     PROPOSITION 054 - LEGISLATURE. LEGISLATION AND...
15     PROPOSITION 055 - TAX EXTENSION TO FUND EDUCAT...
16     PROPOSITION 055 - TAX EXTENSION TO FUND EDUCAT...
17     PROPOSITION 055 - TAX EX

In [125]:
props.prop_name.value_counts().reset_index()

index  prop_name
0   PROPOSITION 057 - CRIMINAL SENTENCES. JUVENILE...         13
1   PROPOSITION 056 - CIGARETTE TAX TO FUND HEALTH...         12
2   PROPOSITION 064- MARIJUANA LEGALIZATION. INITI...         11
3   PROPOSITION 066- DEATH PENALTY. PROCEDURES. IN...          9
4   PROPOSITION 055 - TAX EXTENSION TO FUND EDUCAT...          8
5   PROPOSITION 067- REFERENDUM TO OVERTURN BAN ON...          7
6   PROPOSITION 062- DEATH PENALTY. INITIATIVE STA...          7
7   PROPOSITION 059- SB 254 (CHAPTER 20, STATUTES ...          6
8   PROPOSITION 063- FIREARMS. AMMUNITION SALES. I...          4
9   PROPOSITION 058 - SB 1174 (CHAPTER 753, STATUT...          4
10  PROPOSITION 053 - REVENUE BONDS. STATEWIDE VOT...          4
11  PROPOSITION 054 - LEGISLATURE. LEGISLATION AND...          4
12  PROPOSITION 051 - SCHOOL BONDS. FUNDING FOR K-...          4
13  PROPOSITION 061- STATE PRESCRIPTION DRUG PURCH...          3
14  PROPOSITION 052 - STATE FEES ON HOSPITALS. FED...          3
15  PROPOSITION 060- ADULT FILMS. CONDOMS. HEALTH ...          2
16  PROPOSITION 065- CARRY-OUT BAGS. CHARGES. INIT...          1

In [126]:
props.prop_name.value_counts()

PROPOSITION 057 - CRIMINAL SENTENCES. JUVENILE CRIMINAL PROCEEDINGS AND SENTENCING. INITIATIVE CONSTITUTIONAL AMENDMENT AND STATUTE.                           13
PROPOSITION 056 - CIGARETTE TAX TO FUND HEALTHCARE, TOBACCO USE PREVENTION, RESEARCH, AND LAW ENFORCEMENT. INITIATIVE CONSTITUTIONAL AMENDMENT AND STATUTE.    12
PROPOSITION 064- MARIJUANA LEGALIZATION. INITIATIVE STATUTE.                                                                                                   11
PROPOSITION 066- DEATH PENALTY. PROCEDURES. INITIATIVE STATUTE.                                                                                                 9
PROPOSITION 055 - TAX EXTENSION TO FUND EDUCATION AND HEALTHCARE. INITIATIVE CONSTITUTIONAL AMENDMENT.                                                          8
PROPOSITION 067- REFERENDUM TO OVERTURN BAN ON SINGLE-USE PLASTIC BAGS.                                                                                         7
PROPOSITION 062- DEATH PENAL

In [127]:
weed_prop = props[props.prop_name == "PROPOSITION 064- MARIJUANA LEGALIZATION. INITIATIVE STATUTE."]

In [128]:
weed_prop

ocd_prop_id  calaccess_prop_id  \
74  ocd-contest/232cb72c-1b11-4293-b4e2-a181e80d172e            1381868   
75  ocd-contest/232cb72c-1b11-4293-b4e2-a181e80d172e            1381868   
76  ocd-contest/232cb72c-1b11-4293-b4e2-a181e80d172e            1381868   
77  ocd-contest/232cb72c-1b11-4293-b4e2-a181e80d172e            1381868   
78  ocd-contest/232cb72c-1b11-4293-b4e2-a181e80d172e            1381868   
79  ocd-contest/232cb72c-1b11-4293-b4e2-a181e80d172e            1381868   
80  ocd-contest/232cb72c-1b11-4293-b4e2-a181e80d172e            1381868   
81  ocd-contest/232cb72c-1b11-4293-b4e2-a181e80d172e            1381868   
82  ocd-contest/232cb72c-1b11-4293-b4e2-a181e80d172e            1381868   
83  ocd-contest/232cb72c-1b11-4293-b4e2-a181e80d172e            1381868   
84  ocd-contest/232cb72c-1b11-4293-b4e2-a181e80d172e            1381868   

    ccdc_prop_id                                          prop_name  \
74            83  PROPOSITION 064- MARIJUANA LEGALIZATION. INITI...   
75            83  PROPOSITION 064- MARIJUANA LEGALIZATION. INITI...   
76            83  PROPOSITION 064- MARIJUANA LEGALIZATION. INITI...   
77            83  PROPOSITION 064- MARIJUANA LEGALIZATION. INITI...   
78            83  PROPOSITION 064- MARIJUANA LEGALIZATION. INITI...   
79            83  PROPOSITION 064- MARIJUANA LEGALIZATION. INITI...   
80            83  PROPOSITION 064- MARIJUANA LEGALIZATION. INITI...   
81            83  PROPOSITION 064- MARIJUANA LEGALIZATION. INITI...   
82            83  PROPOSITION 064- MARIJUANA LEGALIZATION. INITI...   
83            83  PROPOSITION 064- MARIJUANA LEGALIZATION. INITI...   
84            83  PROPOSITION 064- MARIJUANA LEGALIZATION. INITI...   

    ccdc_committee_id  calaccess_committee_id  \
74                456                 1381808   
75                457                 1382525   
76                458                 1376077   
77                459                 1382568   
78                460                 1371855   
79                461                 1343793   
80                462                 1386560   
81                463                 1387789   
82                464                 1385506   
83                465                 1385745   
84                466                 1390962   

                                       committee_name committee_position  
74  YES ON 64, CALIFORNIANS TO CONTROL, REGULATE A...            SUPPORT  
75                             NEW APPROACH PAC (MPO)            SUPPORT  
76  CALIFORNIANS FOR SENSIBLE REFORM, SPONSORED BY...            SUPPORT  
77  PUBLIC AND MENTAL HEALTH ADVOCATES AGAINST 64,...             OPPOSE  
78  MARIJUANA POLICY PROJECT OF CALIFORNIA, YES ON 64            SUPPORT  
79  CALIFORNIANS FOR RESPONSIBLE MARIJUANA REFORM,...            SUPPORT  
80         ADULT USE CAMPAIGN FOR PROPOSITION 64; THE            SUPPORT  
81  SAM ACTION, INC., A COMMITTEE AGAINST PROPOSIT...             OPPOSE  
82  DRUG POLICY ACTION - NON PROFIT 501C4, YES ON ...            SUPPORT  
83       FUND FOR POLICY REFORM (NONPROFIT 501(C)(4))            SUPPORT  
84                        MOMS STRONG AGAINST PROP 64             OPPOSE

In [129]:
weed_prop.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 11 entries, 74 to 84
Data columns (total 8 columns):
ocd_prop_id               11 non-null object
calaccess_prop_id         11 non-null int64
ccdc_prop_id              11 non-null int64
prop_name                 11 non-null object
ccdc_committee_id         11 non-null int64
calaccess_committee_id    11 non-null int64
committee_name            11 non-null object
committee_position        11 non-null object
dtypes: int64(4), object(4)
memory usage: 792.0+ bytes


In [130]:
weed_prop_support = weed_prop[weed_prop.committee_position == "SUPPORT"]

In [131]:
weed_prop_oppose = weed_prop[weed_prop.committee_position == "OPPOSE"]

In [132]:
weed_prop_support

ocd_prop_id  calaccess_prop_id  \
74  ocd-contest/232cb72c-1b11-4293-b4e2-a181e80d172e            1381868   
75  ocd-contest/232cb72c-1b11-4293-b4e2-a181e80d172e            1381868   
76  ocd-contest/232cb72c-1b11-4293-b4e2-a181e80d172e            1381868   
78  ocd-contest/232cb72c-1b11-4293-b4e2-a181e80d172e            1381868   
79  ocd-contest/232cb72c-1b11-4293-b4e2-a181e80d172e            1381868   
80  ocd-contest/232cb72c-1b11-4293-b4e2-a181e80d172e            1381868   
82  ocd-contest/232cb72c-1b11-4293-b4e2-a181e80d172e            1381868   
83  ocd-contest/232cb72c-1b11-4293-b4e2-a181e80d172e            1381868   

    ccdc_prop_id                                          prop_name  \
74            83  PROPOSITION 064- MARIJUANA LEGALIZATION. INITI...   
75            83  PROPOSITION 064- MARIJUANA LEGALIZATION. INITI...   
76            83  PROPOSITION 064- MARIJUANA LEGALIZATION. INITI...   
78            83  PROPOSITION 064- MARIJUANA LEGALIZATION. INITI...   
79            83  PROPOSITION 064- MARIJUANA LEGALIZATION. INITI...   
80            83  PROPOSITION 064- MARIJUANA LEGALIZATION. INITI...   
82            83  PROPOSITION 064- MARIJUANA LEGALIZATION. INITI...   
83            83  PROPOSITION 064- MARIJUANA LEGALIZATION. INITI...   

    ccdc_committee_id  calaccess_committee_id  \
74                456                 1381808   
75                457                 1382525   
76                458                 1376077   
78                460                 1371855   
79                461                 1343793   
80                462                 1386560   
82                464                 1385506   
83                465                 1385745   

                                       committee_name committee_position  
74  YES ON 64, CALIFORNIANS TO CONTROL, REGULATE A...            SUPPORT  
75                             NEW APPROACH PAC (MPO)            SUPPORT  
76  CALIFORNIANS FOR SENSIBLE REFORM, SPONSORED BY...            SUPPORT  
78  MARIJUANA POLICY PROJECT OF CALIFORNIA, YES ON 64            SUPPORT  
79  CALIFORNIANS FOR RESPONSIBLE MARIJUANA REFORM,...            SUPPORT  
80         ADULT USE CAMPAIGN FOR PROPOSITION 64; THE            SUPPORT  
82  DRUG POLICY ACTION - NON PROFIT 501C4, YES ON ...            SUPPORT  
83       FUND FOR POLICY REFORM (NONPROFIT 501(C)(4))            SUPPORT

In [133]:
weed_prop_oppose

ocd_prop_id  calaccess_prop_id  \
77  ocd-contest/232cb72c-1b11-4293-b4e2-a181e80d172e            1381868   
81  ocd-contest/232cb72c-1b11-4293-b4e2-a181e80d172e            1381868   
84  ocd-contest/232cb72c-1b11-4293-b4e2-a181e80d172e            1381868   

    ccdc_prop_id                                          prop_name  \
77            83  PROPOSITION 064- MARIJUANA LEGALIZATION. INITI...   
81            83  PROPOSITION 064- MARIJUANA LEGALIZATION. INITI...   
84            83  PROPOSITION 064- MARIJUANA LEGALIZATION. INITI...   

    ccdc_committee_id  calaccess_committee_id  \
77                459                 1382568   
81                463                 1387789   
84                466                 1390962   

                                       committee_name committee_position  
77  PUBLIC AND MENTAL HEALTH ADVOCATES AGAINST 64,...             OPPOSE  
81  SAM ACTION, INC., A COMMITTEE AGAINST PROPOSIT...             OPPOSE  
84                        MOMS STRONG AGAINST PROP 64             OPPOSE

In [134]:
print("Prop. 64 had", weed_prop_support.committee_position.count(), "supporting committees", "compared to", + weed_prop_oppose.committee_position.count(), "opposers.")

Prop. 64 had 8 supporting committees compared to 3 opposers.


In [135]:
weed_prop.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 11 entries, 74 to 84
Data columns (total 8 columns):
ocd_prop_id               11 non-null object
calaccess_prop_id         11 non-null int64
ccdc_prop_id              11 non-null int64
prop_name                 11 non-null object
ccdc_committee_id         11 non-null int64
calaccess_committee_id    11 non-null int64
committee_name            11 non-null object
committee_position        11 non-null object
dtypes: int64(4), object(4)
memory usage: 792.0+ bytes


In [136]:
contrib.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 56379 entries, 0 to 56378
Data columns (total 13 columns):
calaccess_committee_id          56379 non-null int64
committee_name                  56379 non-null object
calaccess_filing_id             56379 non-null int64
date_received                   56379 non-null object
contributor_lastname            56375 non-null object
contributor_firstname           53411 non-null object
contributor_city                56369 non-null object
contributor_state               56363 non-null object
contributor_zip                 56366 non-null object
contributor_employer            48572 non-null object
contributor_occupation          53273 non-null object
contributor_is_self_employed    56379 non-null bool
amount                          56379 non-null float64
dtypes: bool(1), float64(1), int64(2), object(9)
memory usage: 5.2+ MB


# Can I write notes?

# Yes; yes I can.

We can use the merge() method to combine to dataframes into one.

The dataframes must have some type of commmonality with which they can be combined. This called a "key".

In relational databases, when combining two dataframes, the smaller one will be incorporated into the 
larger one, eliminating all the data in the larger one that does not correlate to the smaller one. In this 
way, the smaller dataset acts as a "filter" of sorts.

In our example, we will see how many of the 56,379 contributions (contrib) are connected to Proposition 64
committees (weed_prop). The contributions will be filtered down based on how many of them were to Prop 64
committees. The commonality between them ("key") is the calaccess_committee_id column. We will set them to
a new variable (weed_contribs) in order to save the new dataframe.

The merge() function first takes as arguments the dataframes being used, followed by optional inputs. In this
case, we will use the "on" argument, which allows us to specify the "key" (column) to use for joining.

In [137]:
pd.merge(contrib, weed_prop, on = "calaccess_committee_id")

calaccess_committee_id  \
0                   1386560   
1                   1386560   
2                   1386560   
3                   1386560   
4                   1386560   
5                   1386560   
6                   1386560   
7                   1386560   
8                   1386560   
9                   1386560   
10                  1386560   
11                  1386560   
12                  1386560   
13                  1386560   
14                  1386560   
15                  1386560   
16                  1386560   
17                  1386560   
18                  1386560   
19                  1386560   
20                  1386560   
21                  1386560   
22                  1386560   
23                  1386560   
24                  1386560   
25                  1386560   
26                  1343793   
27                  1343793   
28                  1343793   
29                  1343793   
..                      ...   
830                 1381808   
831                 1381808   
832                 1381808   
833                 1381808   
834                 1381808   
835                 1381808   
836                 1381808   
837                 1381808   
838                 1381808   
839                 1381808   
840                 1381808   
841                 1381808   
842                 1381808   
843                 1381808   
844                 1381808   
845                 1381808   
846                 1381808   
847                 1381808   
848                 1381808   
849                 1381808   
850                 1381808   
851                 1381808   
852                 1381808   
853                 1381808   
854                 1381808   
855                 1381808   
856                 1381808   
857                 1381808   
858                 1381808   
859                 1381808   

                                      committee_name_x  calaccess_filing_id  \
0           ADULT USE CAMPAIGN FOR PROPOSITION 64; THE              2083796   
1           ADULT USE CAMPAIGN FOR PROPOSITION 64; THE              2083796   
2           ADULT USE CAMPAIGN FOR PROPOSITION 64; THE              2083796   
3           ADULT USE CAMPAIGN FOR PROPOSITION 64; THE              2083796   
4           ADULT USE CAMPAIGN FOR PROPOSITION 64; THE              2083796   
5           ADULT USE CAMPAIGN FOR PROPOSITION 64; THE              2083796   
6           ADULT USE CAMPAIGN FOR PROPOSITION 64; THE              2083796   
7           ADULT USE CAMPAIGN FOR PROPOSITION 64; THE              2083796   
8           ADULT USE CAMPAIGN FOR PROPOSITION 64; THE              2083796   
9           ADULT USE CAMPAIGN FOR PROPOSITION 64; THE              2083796   
10          ADULT USE CAMPAIGN FOR PROPOSITION 64; THE              2083796   
11          ADULT USE CAMPAIGN FOR PROPOSITION 64; THE              2083796   
12          ADULT USE CAMPAIGN FOR PROPOSITION 64; THE              2083796   
13          ADULT USE CAMPAIGN FOR PROPOSITION 64; THE              2083796   
14          ADULT USE CAMPAIGN FOR PROPOSITION 64; THE              2083796   
15          ADULT USE CAMPAIGN FOR PROPOSITION 64; THE              2083796   
16          ADULT USE CAMPAIGN FOR PROPOSITION 64; THE              2083796   
17          ADULT USE CAMPAIGN FOR PROPOSITION 64; THE              2098758   
18          ADULT USE CAMPAIGN FOR PROPOSITION 64; THE              2098758   
19          ADULT USE CAMPAIGN FOR PROPOSITION 64; THE              2098758   
20          ADULT USE CAMPAIGN FOR PROPOSITION 64; THE              2131665   
21          ADULT USE CAMPAIGN FOR PROPOSITION 64; THE              2131665   
22          ADULT USE CAMPAIGN FOR PROPOSITION 64; THE              2131665   
23          ADULT USE CAMPAIGN FOR PROPOSITION 64; THE              2131665   
24          ADULT USE CAMPAIGN FOR PROPOSITION 64; THE              2131665   
25          ADULT USE CAMPAI

In [138]:
weed_contribs = pd.merge(contrib, weed_prop, on = "calaccess_committee_id")

In [139]:
weed_contribs.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 860 entries, 0 to 859
Data columns (total 20 columns):
calaccess_committee_id          860 non-null int64
committee_name_x                860 non-null object
calaccess_filing_id             860 non-null int64
date_received                   860 non-null object
contributor_lastname            860 non-null object
contributor_firstname           750 non-null object
contributor_city                860 non-null object
contributor_state               860 non-null object
contributor_zip                 860 non-null object
contributor_employer            657 non-null object
contributor_occupation          742 non-null object
contributor_is_self_employed    860 non-null bool
amount                          860 non-null float64
ocd_prop_id                     860 non-null object
calaccess_prop_id               860 non-null int64
ccdc_prop_id                    860 non-null int64
prop_name                       860 non-null object
ccdc_committee_i

In [140]:
# Out of the the 56,379 total contributions, 860 of them were to committees involved with Prop 64.
# In addition, the new combined dataframe (weed_contribs) has 20 columns, as opposed to the 13 from contrib
# and 8 from weed_prop. calaccess_committee_id column is only shown once, meaning either contrib only
# puts in 12 columns, or weed_prop only puts in 7. Since both dataframes had a committee_name column, they
# have been differentiated in the new dataframe by adding x's and y's to the end of their names (committee_name_x
# and committee_name_y)

In [141]:
supporters = weed_contribs[weed_contribs.committee_position == "SUPPORT"]

In [142]:
weed_contribs_position = weed_contribs.committee_position.value_counts()

In [143]:
weed_contribs_position

SUPPORT    762
OPPOSE      98
Name: committee_position, dtype: int64

In [144]:
supporters.amount.count()

762

In [145]:
opposers = weed_contribs[weed_contribs.committee_position == "OPPOSE"]

In [146]:
opposers.amount.count()

98

# Out of the 860 contributors, 762 supported Prop 64 while 98 opposed it.  

In [147]:
weed_contribs.amount.sum()

35177017.64

In [148]:
supporters.amount.sum()

32675806.0

In [149]:
opposers.amount.sum()

2501211.64

In [150]:
(supporters.amount.sum() / weed_contribs.amount.sum()) * 100

92.88964270479867

In [151]:
(opposers.amount.sum() / weed_contribs.amount.sum()) * 100

7.110357295201333

# The 762 supporters of Prop 64 contributed 32,675,806USD (92.9% of total) compared to the opposers' 2,501,211.64USD (7.1%)...
# ...for a grand total of 35,177,017.64USD in donations.

In [152]:
supporters.amount.median()

100.0

In [153]:
supporters.amount.max()

4000000.0

In [154]:
opposers.amount.median()

658.5

In [155]:
opposers.amount.max()

1000000.0

In [156]:
# The median amount donated by supporters was $100, with the highest donation being $4,000,000; the median doantion
# from opposers was $658.5, with a highest donation of $1,000,000.

In [157]:
max_support_contrib = supporters[supporters.amount == supporters.amount.max()]

In [158]:
max_support_contrib

calaccess_committee_id  \
752                 1381808   

                                      committee_name_x  calaccess_filing_id  \
752  YES ON 64, CALIFORNIANS TO CONTROL, REGULATE A...              2083528   

    date_received            contributor_lastname contributor_firstname  \
752    2016-09-15  PARKER AND AFFILIATED ENTITIES                  SEAN   

    contributor_city contributor_state contributor_zip contributor_employer  \
752        PALO ALTO                CA           94301                  NaN   

    contributor_occupation  contributor_is_self_employed     amount  \
752                    NaN                         False  4000000.0   

                                          ocd_prop_id  calaccess_prop_id  \
752  ocd-contest/232cb72c-1b11-4293-b4e2-a181e80d172e            1381868   

     ccdc_prop_id                                          prop_name  \
752            83  PROPOSITION 064- MARIJUANA LEGALIZATION. INITI...   

     ccdc_committee_id                                   committee_name_y  \
752                456  YES ON 64, CALIFORNIANS TO CONTROL, REGULATE A...   

    committee_position  
752            SUPPORT

In [159]:
max_oppose_contrib = opposers[opposers.amount == opposers.amount.max()]

In [160]:
max_oppose_contrib

calaccess_committee_id  \
608                 1387789   

                                      committee_name_x  calaccess_filing_id  \
608  SAM ACTION, INC., A COMMITTEE AGAINST PROPOSIT...              2083142   

    date_received contributor_lastname contributor_firstname contributor_city  \
608    2016-07-20              SCHAUER                 JULIE             OAKS   

    contributor_state contributor_zip contributor_employer  \
608                PA           19456                  NaN   

    contributor_occupation  contributor_is_self_employed     amount  \
608                RETIRED                         False  1000000.0   

                                          ocd_prop_id  calaccess_prop_id  \
608  ocd-contest/232cb72c-1b11-4293-b4e2-a181e80d172e            1381868   

     ccdc_prop_id                                          prop_name  \
608            83  PROPOSITION 064- MARIJUANA LEGALIZATION. INITI...   

     ccdc_committee_id                                   committee_name_y  \
608                463  SAM ACTION, INC., A COMMITTEE AGAINST PROPOSIT...   

    committee_position  
608             OPPOSE

In [161]:
# The single highest contribution amongst supporters of Prop 64 was by Sean Parker and Affiliated Entities ($4,000,000), while the highest
# donation for Prop 64's opposers was Julie Schauer ($1,000,000).

In [162]:
# The sort_values() function arranges the specified column in an ordered series based on either an ascending or descending
# order. the chosen column is inserted as a string, while the order is described in a boolean manner based on whether or
# not it is ascending.

# Methods like head() can be appended to shorten the list to the top 5, 10, etc. 

In [163]:
# The top 10 supporting donations for Prop 64

In [164]:
supporters.sort_values("amount", ascending = False).head(10)

calaccess_committee_id  \
752                 1381808   
689                 1381808   
458                 1385745   
453                 1385506   
459                 1385745   
751                 1381808   
457                 1385745   
446                 1376077   
753                 1381808   
452                 1385506   

                                      committee_name_x  calaccess_filing_id  \
752  YES ON 64, CALIFORNIANS TO CONTROL, REGULATE A...              2083528   
689  YES ON 64, CALIFORNIANS TO CONTROL, REGULATE A...              2083528   
458       FUND FOR POLICY REFORM (NONPROFIT 501(C)(4))              2082414   
453  DRUG POLICY ACTION - NON PROFIT 501C4, YES ON ...              2083608   
459       FUND FOR POLICY REFORM (NONPROFIT 501(C)(4))              2096507   
751  YES ON 64, CALIFORNIANS TO CONTROL, REGULATE A...              2083528   
457       FUND FOR POLICY REFORM (NONPROFIT 501(C)(4))              2082414   
446  CALIFORNIANS FOR SENSIBLE REFORM, SPONSORED BY...              1982136   
753  YES ON 64, CALIFORNIANS TO CONTROL, REGULATE A...              2083528   
452  DRUG POLICY ACTION - NON PROFIT 501C4, YES ON ...              2083608   

    date_received                               contributor_lastname  \
752    2016-09-15                     PARKER AND AFFILIATED ENTITIES   
689    2016-09-15  DRUG POLICY ACTION - NON PROFIT 501C4, YES ON ...   
458    2016-09-15                             FUND FOR POLICY REFORM   
453    2016-09-15      FUND FOR POLICY REFORM (NONPROFIT 501 (C)(4))   
459    2016-09-27                             FUND FOR POLICY REFORM   
751    2016-07-01                     PARKER AND AFFILIATED ENTITIES   
457    2016-07-07                             FUND FOR POLICY REFORM   
446    2015-04-16           GHOST MANAGEMENT GROUP, LLC DBA WEEDMAPS   
753    2016-09-16                     PARKER AND AFFILIATED ENTITIES   
452    2016-07-07      FUND FOR POLICY REFORM (NONPROFIT 501 (C)(4))   

    contributor_firstname contributor_city contributor_state contributor_zip  \
752                  SEAN        PALO ALTO                CA           94301   
689                   NaN       SACRAMENTO                CA           95815   
458                   NaN       WILMINGTON                DE           19809   
453                   NaN         NEW YORK                NY           10019   
459                   NaN       WILMINGTON                DE           19809   
751                  SEAN        PALO ALTO                CA           94301   
457                   NaN       WILMINGTON                DE           19809   
446                   NaN           IRVINE                CA           92618   
753                  SEAN        PALO ALTO                CA           94301   
452                   NaN         NEW YORK                NY           10019   

    contributor_employer contributor_occupation  contributor_is_self_employed  \
752                  NaN                    NaN                         False   
689                  NaN                    NaN                         False   
458                  NaN                    NaN                         False   
453                  NaN                    NaN                         False   
459                  NaN                    NaN                         False   
751                  NaN                    NaN                         False   
457                  NaN                    NaN                         False   
446                  NaN                    NaN                         False   
753                  NaN                    NaN                         False   
452                  NaN                    NaN                         False   

        amount                                       ocd_prop_id  \
752  4000000.0  ocd-contest/232cb72c-1b11-4293-b4e2-a181e80d172e   
689  2000000.0  ocd-contest/232cb72c-1b11-4293-b4e2-a181e80d172e   
458  1970000.0  ocd-con

In [165]:
# The top 10 opposing donations for Prop 64

In [166]:
opposers.sort_values("amount", ascending = False).head(10)

calaccess_committee_id  \
608                 1387789   
567                 1382568   
565                 1382568   
563                 1382568   
609                 1387789   
606                 1387789   
607                 1387789   
542                 1382568   
605                 1387789   
559                 1382568   

                                      committee_name_x  calaccess_filing_id  \
608  SAM ACTION, INC., A COMMITTEE AGAINST PROPOSIT...              2083142   
567  PUBLIC AND MENTAL HEALTH ADVOCATES AGAINST 64,...              2098048   
565  PUBLIC AND MENTAL HEALTH ADVOCATES AGAINST 64,...              2083071   
563  PUBLIC AND MENTAL HEALTH ADVOCATES AGAINST 64,...              2083071   
609  SAM ACTION, INC., A COMMITTEE AGAINST PROPOSIT...              2083142   
606  SAM ACTION, INC., A COMMITTEE AGAINST PROPOSIT...              2064155   
607  SAM ACTION, INC., A COMMITTEE AGAINST PROPOSIT...              2064155   
542  PUBLIC AND MENTAL HEALTH ADVOCATES AGAINST 64,...              2064127   
605  SAM ACTION, INC., A COMMITTEE AGAINST PROPOSIT...              2064155   
559  PUBLIC AND MENTAL HEALTH ADVOCATES AGAINST 64,...              2083071   

    date_received                               contributor_lastname  \
608    2016-07-20                                            SCHAUER   
567    2016-09-27                                   SAM ACTION, INC.   
565    2016-09-15                                   SAM ACTION, INC.   
563    2016-09-12                                   SAM ACTION, INC.   
609    2016-07-25                                            SCHAUER   
606    2016-06-08                                            SCHAUER   
607    2016-06-13                                            SCHAUER   
542    2016-06-13                                   SAM ACTION, INC.   
605    2016-05-20                                            SCHAUER   
559    2016-08-26  PEACE OFFICERS RESEARCH ASSOC. OF CALIF. POLIT...   

    contributor_firstname contributor_city contributor_state contributor_zip  \
608                 JULIE             OAKS                PA           19456   
567                   NaN       ALEXANDRIA                VA           22314   
565                   NaN       ALEXANDRIA                VA           22314   
563                   NaN       ALEXANDRIA                VA           22314   
609                 JULIE             OAKS                CA           19456   
606                 JULIE             OAKS                PA           19456   
607                 JULIE             OAKS                PA           19456   
542                   NaN       ALEXANDRIA                VA           22314   
605                 JULIE             OAKS                PA           19456   
559                   NaN       SACRAMENTO                CA           95834   

    contributor_employer contributor_occupation  contributor_is_self_employed  \
608                  NaN                RETIRED                         False   
567                  NaN                    NaN                         False   
565                  NaN                    NaN                         False   
563                  NaN                    NaN                         False   
609                  NaN                RETIRED                         False   
606                  NaN                RETIRED                         False   
607                  NaN                RETIRED                         False   
542                  NaN                    NaN                         False   
605                  NaN                RETIRED                         False   
559                  NaN                    NaN                         False   

        amount                                       ocd_prop_id  \
608  1000000.0  ocd-contest/232cb72c-1b11-4293-b4e2-a181e80d172e   
567   400000.0  ocd-contest/232cb72c-1b11-4293-b4e2-a181e80d172e   
565   225000.0  ocd-con

In [167]:
# We want to find the highest overall donors, since people could donate more than once, and the single highest donation may
# not reflect who's actually donated the most.

# Step 1: Create a list of all unique donors for both the supporting and opposing sides. Use First/Last Names, 
# Address data amd Employment data to limit the possibly of accidentally combining contributors with similar info.

# Step 2: Iterate through the list, searching for all donations by each unique donor.
# Step 2.1: Sum the donations for each donor into a separate list or dictionary (whichever's better)

# Step 3: Sort the lists by highest and lowest donors. 

# Optional: Factor in the amount of donations each contributor made, and the average amount they donated.

In [168]:
weed_contribs.groupby(["committee_name_x", "committee_position"]).amount.sum().reset_index().sort_values("amount", ascending = False)

committee_name_x committee_position  \
10  YES ON 64, CALIFORNIANS TO CONTROL, REGULATE A...            SUPPORT   
4        FUND FOR POLICY REFORM (NONPROFIT 501(C)(4))            SUPPORT   
3   DRUG POLICY ACTION - NON PROFIT 501C4, YES ON ...            SUPPORT   
7                              NEW APPROACH PAC (MPO)            SUPPORT   
9   SAM ACTION, INC., A COMMITTEE AGAINST PROPOSIT...             OPPOSE   
8   PUBLIC AND MENTAL HEALTH ADVOCATES AGAINST 64,...             OPPOSE   
1   CALIFORNIANS FOR RESPONSIBLE MARIJUANA REFORM,...            SUPPORT   
2   CALIFORNIANS FOR SENSIBLE REFORM, SPONSORED BY...            SUPPORT   
5   MARIJUANA POLICY PROJECT OF CALIFORNIA, YES ON 64            SUPPORT   
0          ADULT USE CAMPAIGN FOR PROPOSITION 64; THE            SUPPORT   
6                         MOMS STRONG AGAINST PROP 64             OPPOSE   

         amount  
10  15734047.00  
4    6140000.00  
3    4470000.00  
7    3865000.00  
9    1364000.00  
8    1130930.00  
1     860470.00  
2     850000.00  
5     739200.00  
0      17089.00  
6       6281.64

In [169]:
top_donors = weed_contribs.groupby(["contributor_firstname","contributor_lastname","committee_position"]).amount.sum().reset_index().sort_values("amount", ascending = False)

In [170]:
top_donors

contributor_firstname            contributor_lastname committee_position  \
483                  SEAN  PARKER AND AFFILIATED ENTITIES            SUPPORT   
482                  SEAN                          PARKER            SUPPORT   
281                 JULIE                         SCHAUER             OPPOSE   
108                DANIEL                           LEWIS            SUPPORT   
194                 HENRY                   VAN AMERINGEN            SUPPORT   
405              NICHOLAS                        PRITZKER            SUPPORT   
421                 PETER                           THIEL            SUPPORT   
456                 ROGER                         JENKINS            SUPPORT   
23                 ANGELA                          HOWARD            SUPPORT   
322              LAWRENCE                            HESS            SUPPORT   
221                 JASON                            POST            SUPPORT   
460              ROGER B.                         MCNAMEE            SUPPORT   
548           ZACHARY PYM                        WILLIAMS            SUPPORT   
239                JEREMY                         ZACHARY            SUPPORT   
440                ROBERT                           ADAMS             OPPOSE   
320                LAURIE                        BENENSON            SUPPORT   
63                CAMILLA                           FIELD            SUPPORT   
375               MICHAEL                            DEAN             OPPOSE   
361            MARY QUINN                         DELANEY            SUPPORT   
488             SHARON K.                        CHRISTIE            SUPPORT   
515                 TERRY                          TURNER            SUPPORT   
301              KHURSHID                           KHOJA            SUPPORT   
277                  JULI                        CROCKETT            SUPPORT   
149                  DOUG                        LANPHERE            SUPPORT   
76                CHARLES                           ROVEN            SUPPORT   
158                 EMILY                          PAXHIA            SUPPORT   
414               PATRICK                       MCMANAMON            SUPPORT   
507                STEVEN                          PERSKY            SUPPORT   
382               MICHAEL                       STEINMETZ            SUPPORT   
377               MICHAEL                           KLEIN            SUPPORT   
..                    ...                             ...                ...   
75                CHARLES                            LOVE            SUPPORT   
191              GRETCHEN                         BERGMAN            SUPPORT   
387              MICHELLE                           WOLFF            SUPPORT   
385              MICHELLE                          SEXTON            SUPPORT   
189                  GREG                           SMITH            SUPPORT   
389                  MIKE                         BARNETT            SUPPORT   
534                 WENDY                          COLTON            SUPPORT   
168                 ETHAN                         BROOKES            SUPPORT   
85           CHRISTINE A.                          PORTER            SUPPORT   
459                 ROGER                           TOWER            SUPPORT   
492                  SKIP                          ROACHE            SUPPORT   
335              LJUBOMIR                       BUTUROVIC            SUPPORT   
418                 PETER                          KIRSCH            SUPPORT   
326               LEONARD                        KAUFMANN            SUPPORT   
104                  DANA                          HIRSCH            SUPPORT   
105                DANIEL                          DODSON            SUPPORT   
125                 DAVID                          MURRAY            SUPPORT   
272                  JOSH                       KLEYMEYER            SUPPORT   
51  

In [171]:
    top_supporters = supporters.groupby(["contributor_firstname","contributor_lastname","committee_position"]).amount.sum().reset_index().sort_values("amount", ascending = False).head(10)

In [172]:
top_supporters

contributor_firstname            contributor_lastname committee_position  \
444                  SEAN  PARKER AND AFFILIATED ENTITIES            SUPPORT   
443                  SEAN                          PARKER            SUPPORT   
102                DANIEL                           LEWIS            SUPPORT   
180                 HENRY                   VAN AMERINGEN            SUPPORT   
374              NICHOLAS                        PRITZKER            SUPPORT   
390                 PETER                           THIEL            SUPPORT   
421                 ROGER                         JENKINS            SUPPORT   
22                 ANGELA                          HOWARD            SUPPORT   
297              LAWRENCE                            HESS            SUPPORT   
425              ROGER B.                         MCNAMEE            SUPPORT   

        amount  
444  7250000.0  
443  1500000.0  
102  1250000.0  
180  1000000.0  
374   900000.0  
390   300000.0  
421    75000.0  
22     65000.0  
297    30000.0  
425    10000.0

In [173]:
top_opposers = opposers.groupby(["contributor_firstname","contributor_lastname","committee_position"]).amount.sum().reset_index().sort_values("amount", ascending = False).head(10)

In [174]:
top_opposers

contributor_firstname contributor_lastname committee_position     amount
20                 JULIE              SCHAUER             OPPOSE  1364000.0
31                ROBERT                ADAMS             OPPOSE     6500.0
29               MICHAEL                 DEAN             OPPOSE     5000.0
13                GEORGE                 BELL             OPPOSE     1550.0
12               FARSHAD               FARDAD             OPPOSE     1000.0
17               JESSICA               MACIEL             OPPOSE     1000.0
9                  EDITH              REUVENI             OPPOSE     1000.0
3              CHRISTINE               MILLER             OPPOSE      871.0
25                  LORI             ROBINSON             OPPOSE      817.0
8                 DWIGHT                 BELL             OPPOSE      730.0

In [175]:
weed_contribs[weed_contribs.contributor_state != "CA"].amount.sum()

17685824.64

In [176]:
# Contributions from outside the state of California

In [177]:
weed_contribs[weed_contribs.contributor_state == "CA"].amount.sum()

17491193.0

In [178]:
# Contributions from within the state of California

In [179]:
weed_contribs.amount.sum()

35177017.64

In [180]:
(weed_contribs[weed_contribs.contributor_state != "CA"].amount.sum() / weed_contribs.amount.sum()) * 100

50.2766460221157

In [181]:
# Percentage of total donations that came from outside California

In [182]:
(weed_contribs[weed_contribs.contributor_state == "CA"].amount.sum() / weed_contribs.amount.sum()) * 100

49.723353977884294

In [183]:
# Percentage of total donations that came from within California

In [184]:
# The contributions from within the state of California only slightly edged out those from out of state (50.28% vs 49.72%).

In [185]:
top_employers = weed_contribs.groupby("contributor_employer").amount.sum().reset_index().sort_values("amount", ascending = False).head(10)

In [186]:
top_employers

contributor_employer     amount
347     SEAN PARKER FOUNDATION  1250000.0
318                    RETIRED   503671.0
377   TAO CAPITAL PARTNERS LLC   450000.0
378  TAO CAPITAL PARTNERS, LLC   450000.0
388         THIEL CAPITAL, LLC   300000.0
346   SEAN N PARKER FOUNDATION   250000.0
291                       NONE    88427.0
134                   DESIGNER    65000.0
237       LEHBROS LIMITED, LLC    30000.0
152         ELEVATION PARTNERS    10000.0

In [187]:
# The top 10 employers of donors by amount donated

In [188]:
top_occupations = weed_contribs.groupby("contributor_occupation").amount.sum().reset_index().sort_values("amount", ascending = False).head(10)

In [189]:
top_occupations

contributor_occupation     amount
177                  RETIRED  3128025.0
34                  CHAIRMAN  2400000.0
147                PRESIDENT   332900.0
109                 INVESTOR    95200.0
61                  DESIGNER    65100.0
36   CHIEF EXECUTIVE OFFICER    13020.0
64                  DIRECTOR    12450.0
203               UNEMPLOYED    11102.0
14                  ATTORNEY    10710.0
135                    OWNER    10180.0

In [190]:
# The top 10 occupations of donors by amount donated

In [191]:
top_support_employer = supporters.groupby("contributor_employer").amount.sum().reset_index().sort_values("amount", ascending = False).head(10)

In [192]:
top_support_employer

contributor_employer     amount
322     SEAN PARKER FOUNDATION  1250000.0
294                    RETIRED   502300.0
351  TAO CAPITAL PARTNERS, LLC   450000.0
350   TAO CAPITAL PARTNERS LLC   450000.0
361         THIEL CAPITAL, LLC   300000.0
321   SEAN N PARKER FOUNDATION   250000.0
268                       NONE    86927.0
115                   DESIGNER    65000.0
216       LEHBROS LIMITED, LLC    30000.0
95         CONDE NAST TRAVELER    10000.0

In [193]:
The top 10 employers of supporting donors by amount donated

SyntaxError: invalid syntax (<ipython-input-193-28b61595dcc0>, line 1)

In [194]:
top_oppose_employer = opposers.groupby("contributor_employer").amount.sum().reset_index().sort_values("amount", ascending = False).head(10)

In [195]:
top_oppose_employer

contributor_employer  amount
17                   CS  6500.0
18                  CSI  5000.0
4     BOSTON SCIENTIFIC  1550.0
23                 NONE  1500.0
25              RETIRED  1371.0
22                  NON  1000.0
19     GLOBALWIDE MEDIA  1000.0
29                SFWKC   817.0
30  STATE OF CALIFORNIA   730.0
20            LA COUNTY   500.0

In [196]:
# The top 10 employers of donors by amount donated

In [197]:
top_support_occupation = supporters.groupby("contributor_occupation").amount.sum().reset_index().sort_values("amount", ascending = False).head(10)

In [198]:
top_support_occupation

contributor_occupation     amount
29                  CHAIRMAN  2400000.0
155                  RETIRED  1763725.0
127                PRESIDENT   332800.0
95                  INVESTOR    95200.0
49                  DESIGNER    65100.0
31   CHIEF EXECUTIVE OFFICER    13020.0
51                  DIRECTOR    12450.0
177               UNEMPLOYED    11102.0
12                  ATTORNEY    10710.0
119                    OWNER    10180.0

In [199]:
# The top 10 occupations of donors by amount donated

In [200]:
top_oppose_occupation = opposers.groupby("contributor_occupation").amount.sum().reset_index().sort_values("amount", ascending = False).head(10)

In [201]:
top_oppose_occupation

contributor_occupation     amount
26                RETIRED  1364300.0
1           ADMINISTRATOR     6500.0
0          ADMINISTRATIVE     5000.0
30       TECHNICAL WRITER     1550.0
27                   SELF     1500.0
14              HOMEMAKER     1100.0
22                  PRES.     1000.0
19                    PHD      871.0
5         CHIEF OF POLICE      850.0
13              DIETITIAN      817.0

In [202]:
# The top 10 occupations of opposers by amount donated 

In [203]:
smallest_committees = weed_contribs.groupby("committee_name_x").amount.sum().reset_index().sort_values("amount", ascending = True).head(5)

In [204]:
smallest_committees

committee_name_x     amount
6                        MOMS STRONG AGAINST PROP 64    6281.64
0         ADULT USE CAMPAIGN FOR PROPOSITION 64; THE   17089.00
5  MARIJUANA POLICY PROJECT OF CALIFORNIA, YES ON 64  739200.00
2  CALIFORNIANS FOR SENSIBLE REFORM, SPONSORED BY...  850000.00
1  CALIFORNIANS FOR RESPONSIBLE MARIJUANA REFORM,...  860470.00

In [205]:
# The top 5 lowest earning committees

In [206]:
smallest_support_committees = supporters.groupby("committee_name_x").amount.sum().reset_index().sort_values("amount", ascending = True).head(5)

In [207]:
smallest_support_committees

committee_name_x     amount
0         ADULT USE CAMPAIGN FOR PROPOSITION 64; THE    17089.0
5  MARIJUANA POLICY PROJECT OF CALIFORNIA, YES ON 64   739200.0
2  CALIFORNIANS FOR SENSIBLE REFORM, SPONSORED BY...   850000.0
1  CALIFORNIANS FOR RESPONSIBLE MARIJUANA REFORM,...   860470.0
6                             NEW APPROACH PAC (MPO)  3865000.0

In [208]:
# The top 5 lowest-earning support committees

In [209]:
smallest_oppose_committees = opposers.groupby("committee_name_x").amount.sum().reset_index().sort_values("amount", ascending = True).head(10)

In [210]:
smallest_oppose_committees

committee_name_x      amount
0                        MOMS STRONG AGAINST PROP 64     6281.64
1  PUBLIC AND MENTAL HEALTH ADVOCATES AGAINST 64,...  1130930.00
2  SAM ACTION, INC., A COMMITTEE AGAINST PROPOSIT...  1364000.00

In [211]:
# The top 3 lowest-earning support committees

In [212]:
largest_support_committees = supporters.groupby("committee_name_x").amount.sum().reset_index().sort_values("amount", ascending = False).head(5)

In [213]:
largest_support_committees

committee_name_x      amount
7  YES ON 64, CALIFORNIANS TO CONTROL, REGULATE A...  15734047.0
4       FUND FOR POLICY REFORM (NONPROFIT 501(C)(4))   6140000.0
3  DRUG POLICY ACTION - NON PROFIT 501C4, YES ON ...   4470000.0
6                             NEW APPROACH PAC (MPO)   3865000.0
1  CALIFORNIANS FOR RESPONSIBLE MARIJUANA REFORM,...    860470.0

In [214]:
# The top 5 largest support committees

In [215]:
largest_oppose_committees = opposers.groupby("committee_name_x").amount.sum().reset_index().sort_values("amount", ascending = False).head(5)

In [216]:
largest_oppose_committees

committee_name_x      amount
2  SAM ACTION, INC., A COMMITTEE AGAINST PROPOSIT...  1364000.00
1  PUBLIC AND MENTAL HEALTH ADVOCATES AGAINST 64,...  1130930.00
0                        MOMS STRONG AGAINST PROP 64     6281.64

In [217]:
# The top 3 largest opposing committees

In [218]:
# Adding a change for Git test.